File: common/src/main/java/cc/turtl/chiselmon/feature/eggspy/EggCache.java
package cc.turtl.chiselmon.feature.eggspy;

import cc.turtl.chiselmon.api.duck.DuckPreviewPokemon;
import com.cobblemon.mod.common.pokemon.Pokemon;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;

import java.util.UUID;
import java.util.concurrent.TimeUnit;

public class EggCache {
    private static final Cache<UUID, EggDummy> EGGS_CACHE = CacheBuilder.newBuilder()
            .maximumSize(200)
            .expireAfterAccess(10, TimeUnit.MINUTES)
            .build();

    /**
     * Returns back the original pokemon, or its egg preview if possible.
     * This function assumes the mod and eggspy feature is enabled, check this before calling!
     */
    public static Pokemon getPreview(Pokemon pokemon) {
        if (pokemon instanceof EggDummy) return pokemon;
        if (!((DuckPreviewPokemon) pokemon).chiselmon$isEgg()) return pokemon;

        UUID uuid = pokemon.getUuid();
        EggDummy cached = EGGS_CACHE.getIfPresent(uuid);

        if (cached != null) {
            return cached;
        }

        // Try to generate, return original if parsing fails
        return EggDummy.from(pokemon)
                .map(dummy -> {
                    EGGS_CACHE.put(uuid, dummy);
                    return (Pokemon) dummy;
                })
                .orElse(pokemon);
    }

    public static void invalidate(UUID uuid) {
        EGGS_CACHE.invalidate(uuid);
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/feature/eggspy/EggCache.kt
package cc.turtl.chiselmon.client.feature.eggspy

import cc.turtl.chiselmon.client.api.duck.DuckPreviewPokemon
import com.cobblemon.mod.common.api.pokemon.feature.IntSpeciesFeature
import com.cobblemon.mod.common.pokemon.Pokemon
import com.google.common.cache.CacheBuilder
import java.util.UUID
import java.util.concurrent.TimeUnit

object EggCache {
    private val cache = CacheBuilder.newBuilder()
        .maximumSize(200)
        .expireAfterAccess(10, TimeUnit.MINUTES)
        .build<UUID, EggDummy>()

    /**
     * Returns the original pokemon, or its egg preview if available.
     * Live progress fields are refreshed from the given pokemon on every call,
     * so the dummy stays accurate without the whole object needing to be invalidated.
     */
    @JvmStatic
    fun getPreview(pokemon: Pokemon): Pokemon {
        if (pokemon is EggDummy) return pokemon
        if (!(pokemon as DuckPreviewPokemon).`chiselmon$isEgg`()) return pokemon

        val uuid = pokemon.uuid
        val dummy = cache.getIfPresent(uuid)
            ?: EggDummy.from(pokemon)?.also { cache.put(uuid, it) }
            ?: return pokemon

        dummy.totalSteps = pokemon.persistentData.getInt("TotalSteps")
        dummy.hatchPercentage = pokemon.getFeature<IntSpeciesFeature>(EggDummy.HATCH_PERCENTAGE_FEATURE)?.value ?: 0

        return dummy
    }

    @JvmStatic
    fun invalidate(uuid: UUID) = cache.invalidate(uuid)
}